services/s3/src/main/java/software/amazon/awssdk/services/s3/multipart/MultipartConfiguration.java

```diff
@@ -0,0 +1,206 @@
+/*
+ * Copyright Amazon.com, Inc. or its affiliates. All Rights Reserved.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License").
+ * You may not use this file except in compliance with the License.
+ * A copy of the License is located at
+ *
+ *  http://aws.amazon.com/apache2.0
+ *
+ * or in the "license" file accompanying this file. This file is distributed
+ * on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing
+ * permissions and limitations under the License.
+ */
+
+package software.amazon.awssdk.services.s3.multipart;
+
+import java.util.function.Consumer;
+import software.amazon.awssdk.annotations.SdkPublicApi;
+import software.amazon.awssdk.core.async.AsyncRequestBody;
+import software.amazon.awssdk.core.async.AsyncResponseTransformer;
+import software.amazon.awssdk.services.s3.S3AsyncClient;
+import software.amazon.awssdk.services.s3.S3AsyncClientBuilder;
+import software.amazon.awssdk.services.s3.model.CopyObjectRequest;
+import software.amazon.awssdk.services.s3.model.GetObjectRequest;
+import software.amazon.awssdk.utils.AttributeMap;
+import software.amazon.awssdk.utils.builder.CopyableBuilder;
+import software.amazon.awssdk.utils.builder.ToCopyableBuilder;
+
+/**
+ * Class that hold configuration properties related to multipart operation for a {@link S3AsyncClient}. Passing this class to the
+ * {@link S3AsyncClientBuilder#multipartConfiguration(MultipartConfiguration)} will enable automatic conversion of
+ * {@link S3AsyncClient#putObject(Consumer, AsyncRequestBody)}, {@link S3AsyncClient#copyObject(CopyObjectRequest)} to their
+ * respective multipart operation.
+ * <p></p>
+ * <em>Note</em>: The multipart operation for {@link S3AsyncClient#getObject(GetObjectRequest, AsyncResponseTransformer)} is
+ * temporarily disabled and will result in throwing a {@link UnsupportedOperationException} if called when configured for
+ * multipart operation.
+ */
+@SdkPublicApi
+public final class MultipartConfiguration implements ToCopyableBuilder<MultipartConfiguration.Builder, MultipartConfiguration> {
+    public static final AttributeMap.Key<MultipartConfiguration> MULTIPART_CONFIGURATION_KEY =
+        new AttributeMap.Key<MultipartConfiguration>(MultipartConfiguration.class){};
+    public static final AttributeMap.Key<Boolean> MULTIPART_ENABLED_KEY =
+        new AttributeMap.Key<Boolean>(Boolean.class){};
```

Comment:
These seem internal. Should they be on an internal class? MultipartS3AsyncClient maybe?

Reply:
Shouldn't they be public so users can retrieve those values with clientContextParams?

Reply:
Do we need users to be able to retrieve those values?

Reply:
We dont *need* them to, no, but they might want to. Is there any good reason we would want to prevent users to access those value? The one I see is it might be confusing if they retrieve the config to change it and it has no effect since these have effect on when the client is build, not for every request. Actually, now that I think of it, that might be reason enough to prevent it actually...

Reply:
Also that it's a public API we'd have to maintain in perpetuity.

Reply:
I'll move them to `S3AsyncClientDecorator` it is an `@SdkInternalApi` class and is where they are used outside of codegen. And where it needs to be used by the codegen already knows about it, so that seems to be a cohesive solution.